src/archive.rs

```diff
@@ -0,0 +1,154 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Tools for creating and inserting into tarballs.
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use camino::Utf8Path;
+use flate2::write::GzEncoder;
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use tar::Builder;
+
+#[async_trait]
+pub trait AsyncAppendFile {
+    async fn append_file_async<P>(&mut self, path: P, file: &mut File) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send;
+
+    async fn append_path_with_name_async<P, N>(&mut self, path: P, name: N) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send,
+        N: AsRef<Utf8Path> + Send;
+
+    async fn append_dir_all_async<P, Q>(&mut self, path: P, src_path: Q) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send,
+        Q: AsRef<Utf8Path> + Send;
+}
+
+#[async_trait]
+impl<W: Encoder> AsyncAppendFile for Builder<W> {
+    async fn append_file_async<P>(&mut self, path: P, file: &mut File) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send,
+    {
+        tokio::task::block_in_place(move || self.append_file(path.as_ref(), file))
+    }
+
+    async fn append_path_with_name_async<P, N>(&mut self, path: P, name: N) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send,
+        N: AsRef<Utf8Path> + Send,
+    {
+        tokio::task::block_in_place(move || {
+            self.append_path_with_name(path.as_ref(), name.as_ref())
+        })
+    }
+
+    async fn append_dir_all_async<P, Q>(&mut self, path: P, src_path: Q) -> std::io::Result<()>
+    where
+        P: AsRef<Utf8Path> + Send,
+        Q: AsRef<Utf8Path> + Send,
+    {
+        tokio::task::block_in_place(move || self.append_dir_all(path.as_ref(), src_path.as_ref()))
+    }
+}
+
+/// Helper to open a tarfile for reading/writing.
+pub fn create_tarfile<P: AsRef<Utf8Path> + std::fmt::Debug>(tarfile: P) -> Result<File> {
+    OpenOptions::new()
+        .write(true)
+        .read(true)
+        .truncate(true)
+        .create(true)
+        .open(tarfile.as_ref())
+        .map_err(|err| anyhow!("Cannot create tarfile {:?}: {}", tarfile, err))
+}
+
+/// Helper to open a tarfile for reading.
+pub fn open_tarfile<P: AsRef<Utf8Path> + std::fmt::Debug>(tarfile: P) -> Result<File> {
+    OpenOptions::new()
+        .read(true)
+        .open(tarfile.as_ref())
+        .map_err(|err| anyhow!("Cannot open tarfile {:?}: {}", tarfile, err))
+}
+
+pub trait Encoder: std::io::Write + Send {}
+impl<T> Encoder for T where T: std::io::Write + Send {}
+
+pub struct ArchiveBuilder<E: Encoder> {
+    pub builder: tar::Builder<E>,
+}
+
+impl<E: Encoder> ArchiveBuilder<E> {
+    pub fn new(builder: tar::Builder<E>) -> Self {
+        Self { builder }
+    }
+
+    pub fn into_inner(self) -> Result<E> {
+        self.builder
+            .into_inner()
+            .with_context(|| "Finalizing archive")
```

Comment:
given this is a static string, `context` is slightly simpler than `.with_context`, but it'll get compiled away to the same thing.

Reply:
Updated across the codebase with 021e08f811036bfcbc75c1262cd2ee5899362ef3